library/common/http/dispatcher.cc

```diff
@@ -229,71 +321,122 @@ envoy_status_t Dispatcher::sendMetadata(envoy_stream_t, envoy_headers) { return
-  case ENVOY_NET_WLAN:
-    // The ASSERT above ensures the cluster_manager_ is safe to access.
-    return TS_UNCHECKED_READ(cluster_manager_)->httpAsyncClientForCluster("base_wlan");
-  case ENVOY_NET_WWAN:
-    return TS_UNCHECKED_READ(cluster_manager_)->httpAsyncClientForCluster("base_wwan");
-  case ENVOY_NET_GENERIC:
-  default:
-    return TS_UNCHECKED_READ(cluster_manager_)->httpAsyncClientForCluster("base");
+    return ENVOY_SUCCESS;
   }
+  return ENVOY_FAILURE;
 }
 
-Dispatcher::DirectStream* Dispatcher::getStream(envoy_stream_t stream) {
-  // The dispatch_lock_ does not need to guard the event_dispatcher_ pointer here because this
-  // function should only be called from the context of Envoy's event dispatcher.
-  ASSERT(TS_UNCHECKED_READ(event_dispatcher_)->isThreadSafe(),
-         "stream interaction must be performed on the event_dispatcher_'s thread.");
+Dispatcher::DirectStreamSharedPtr Dispatcher::getStream(envoy_stream_t stream) {
+  Thread::LockGuard lock(streams_lock_);
+
   auto direct_stream_pair_it = streams_.find(stream);
-  return (direct_stream_pair_it != streams_.end()) ? direct_stream_pair_it->second.get() : nullptr;
+  // Returning will copy the shared_ptr and increase the ref count. Moreover, this is safe because
+  // creation of the return value happens before destruction of local variables:
+  // https://stackoverflow.com/questions/33150508/in-c-which-happens-first-the-copy-of-a-return-object-or-local-objects-destru
+  return (direct_stream_pair_it != streams_.end()) ? direct_stream_pair_it->second : nullptr;
 }
 
 void Dispatcher::cleanup(envoy_stream_t stream_handle) {
-  DirectStream* direct_stream = getStream(stream_handle);
-
+  ASSERT(TS_UNCHECKED_READ(event_dispatcher_)->isThreadSafe(),
+         "stream cleanup must be performed on the event_dispatcher_'s thread.");
+  Dispatcher::DirectStreamSharedPtr direct_stream = getStream(stream_handle);
   RELEASE_ASSERT(direct_stream,
                  "cleanup is a private method that is only called with stream ids that exist");
 
+  // The DirectStream should live through synchronous code that already has a reference to it.
+  // Hence why it is scheduled for deferred deletion. If this was all that was needed then it
+  // would be sufficient to return a shared_ptr in getStream. However, deferred deletion is still
+  // required because in Dispatcher::resetStream the DirectStream needs to live for as long and
+  // the HCM's ActiveStream lives. Hence its deletion needs to live beyond the synchronous code in
+  // Dispatcher::resetStream.
+  // TODO: currently upstream Envoy does not have a deferred delete version for shared_ptr. This
+  // means that instead of efficiently queuing the deletes for one event in the event loop, all
+  // deletes here get queued up as individual posts.
+  TS_UNCHECKED_READ(event_dispatcher_)->post([direct_stream]() -> void {
+    ENVOY_LOG(debug, "[S{}] deferred deletion of stream", direct_stream->stream_handle_);
```

Comment:
Just want to be sure that the optimizer won't optimize away this capture of `direct_stream` - even if `ENVOY_LOG` is compiled away. (Even if this would definitively/arguably be an optimizer bug, I've seen this sort of thing happen before.)

Reply:
I really enjoy your eye for things that might end up being subtle bugs in the future. https://stackoverflow.com/a/12718425 and the comments below state that an explicit capture like above will not be optimized out even if the variable is not used in the lambda body.